### PR TITLE
[Frontend] Add domains and tags to challenges

### PR DESCRIPTION
**Pull request by Suryansh5545**
Added Tags to backend, frontend, updated test case for the same.

Peek 2023-05-27 18-11

### CODE REVIEW

File: apps/challenges/challenge_config_utils.py [hunk 1]
```diff
@@ -506,6 +506,26 @@ def validate_challenge_config_util(
             )
             error_messages.append(message)
 
+        # Verify Tags are limited to 4
+        if "tags" in yaml_file_data:
+            tags_data = yaml_file_data["tags"]
+            if len(tags_data) > 4:
+                message = (
```
Comment (gchhablani) [inline apps/challenges/challenge_config_utils.py:513]: Nice!

File: apps/challenges/challenge_config_utils.py [hunk 2]
```diff
@@ -506,6 +506,26 @@ def validate_challenge_config_util(
...
+                    "ERROR: Tags are limited to 4. Please remove extra tags"
+                    " then try again!"
+                )
+                error_messages.append(message)
+
+        # Verify Domain name is correct
+        if "domain" in yaml_file_data:
+            domain_value = yaml_file_data["domain"]
+            if domain_value not in ["CV", "NLP", "RL"]:
```
Comment (gchhablani) [inline apps/challenges/challenge_config_utils.py:522]: I am not sure if this should be hardcoded.
Is there no way to take this list of domains from somewhere else?
Maybe import from models or somewhere?

File: apps/challenges/serializers.py [hunk 3]
```diff
@@ -15,12 +15,40 @@
...
+
+    class Meta:
+        model = ChallengeTags
+        fields = "__all__"
+
+
 class ChallengeSerializer(serializers.ModelSerializer):
 
     is_active = serializers.ReadOnlyField()
+    list_tags = ChallengeTagsSerializer(many=True)
```
Comment (gchhablani) [inline apps/challenges/serializers.py:46]: Why are we doing this?

File: docker/dev/django/container-start.sh [hunk 4]
```diff
@@ -1,4 +1,5 @@
 #!/bin/sh
+python manage.py makemigrations && \
```
Comment (gchhablani) [inline docker/dev/django/container-start.sh:2]: What is this change for?

File: frontend/src/js/controllers/challengeCtrl.js [hunk 5]
```diff
@@ -1076,6 +1075,11 @@
 
             utilities.sendRequest(parameters);
         };
+
+        if (vm.phaseSplitId) {
```
Comment (gchhablani) [inline frontend/src/js/controllers/challengeCtrl.js:1079]: What is this change for?

File: apps/challenges/challenge_config_utils.py (hunk 2 above)
Comment (Suryansh5545) [reply apps/challenges/challenge_config_utils.py:522]: It is definitely possible,  I will change this part

File: apps/challenges/admin.py [hunk 6]
```diff
@@ -431,3 +432,10 @@ def get_challenge_name_and_id(self, obj):
 
     get_challenge_name_and_id.short_description = "Challenge Name - ID"
     get_challenge_name_and_id.admin_order_field = "challenge_phase__challenge"
+
+
+@admin.register(ChallengeTags)
+class TagsAdmin(ImportExportTimeStampedAdmin):
```
[the remark this answers is not in the document]
Comment (Suryansh5545) [reply apps/challenges/admin.py:438]: Exposes the ChallengeTags model to diango admin ui

File: apps/challenges/models.py [hunk 7]
```diff
@@ -27,6 +27,22 @@ def get_default_eval_metric():
     return ["Accuracy"]
 
 
+class ChallengeTags(models.Model):
```
[the remark this answers is not in the document]
Comment (Suryansh5545) [reply apps/challenges/models.py:30]: Like a ArrayField? We can do that but that won't be good looking on the admin side.

File: apps/challenges/serializers.py (hunk 3 above)
Comment (Suryansh5545) [reply apps/challenges/serializers.py:46]: We get the Id or index of list_tags as it is a manytomany in challenge. So we serialize it to get the value or name of tag

File: apps/challenges/views.py [hunk 8]
```diff
@@ -1403,6 +1404,37 @@ def create_challenge_using_zip_file(request, challenge_host_team_pk):
...
+                # Remove tags not present in the YAML file
+                challenge.list_tags.set(ChallengeTags.objects.filter(tag_name__in=new_tags))
+
+                # Add new tags to the challenge
+                for tag_name in new_tags:
+                    tag_obj, _ = ChallengeTags.objects.get_or_create(tag_name=tag_name, challenge=challenge)
+                    challenge.list_tags.add(tag_obj)
+            else:
+                # Remove all existing tags if no tags are defined in the YAML file
+                challenge.list_tags.clear()
```
[the remark this answers is not in the document]
Comment (Suryansh5545) [reply apps/challenges/views.py:1421]: Yes it assumes that the tags that aren't listed in the updated config send are suppose to be removed

File: docker/dev/django/container-start.sh (hunk 4 above)
Comment (Suryansh5545) [reply docker/dev/django/container-start.sh:2]: Whenever a change is made to django make we have to run the command makemigrations separate every time. Shouldn't we put it in the start so all migrations are made and then migrated to db?

File: frontend/src/css/modules/challenge.scss [hunk 9]
```diff
@@ -60,7 +60,7 @@ a.active-challenge {
 }
 
 .ev-challenge-card {
-    height: 425px;
+    min-height: 460px;
```
[the remark this answers is not in the document]
Comment (Suryansh5545) [reply frontend/src/css/modules/challenge.scss:63]: Adding the tags and domain made the card data get a  bit cut due to small size. It tries to compensate for it. But we still have that effect

File: apps/challenges/views.py [hunk 10]
```diff
@@ -3392,6 +3469,37 @@ def create_or_update_github_challenge(request, challenge_host_team_pk):
...
+                        # Remove tags not present in the YAML file
+                        challenge.list_tags.set(ChallengeTags.objects.filter(tag_name__in=new_tags))
+
+                        # Add new tags to the challenge
+                        for tag_name in new_tags:
+                            tag_obj, _ = ChallengeTags.objects.get_or_create(tag_name=tag_name, challenge=challenge)
+                            challenge.list_tags.add(tag_obj)
+                    else:
+                        # Remove all existing tags if no tags are defined in the YAML file
+                        challenge.list_tags.clear()
```
[the remark this answers is not in the document]
Comment (Suryansh5545) [reply apps/challenges/views.py:3486]: yes it removed the old tag, in new approach too

**Pull request closed without merging**